accounts/models.py

```diff
@@ -186,10 +186,20 @@ def can_create_study(self):
                 self.has_perm(
                     LabPermission.CREATE_LAB_ASSOCIATED_STUDY.codename, obj=lab
                 )
-                for lab in self.labs.values("id")
+                for lab in self.labs.only("id")
             ]
         )
 
+    def labs_user_can_create_study_in(self):
+        lab_ids = [
```

Comment:
`get_objects_for_user` will keep it all in one query rather than evaluating each lab with `self.has_perm`

You probably want something like this instead:

```python
self.labs.filter(id__in=get_objects_for_user(self, LabPermission.CREATE_LAB_ASSOCIATED_STUDY.prefixed_codename).only("id"))
```

As an interesting sidenote... doing your first query (the list comp) in the PyCharm console actually breaks with a `SynchronousOnlyOperation` error. I'm going to pull down your code in a second to test it and see if this is just a weird issue with the PC console, or if there's something else going on.

Reply:
ok, looks like it's just a weird thing with the PyCharm console (the REPL debugger probably uses an async loop, which Django doesn't like). Regardless, I just tested it and confirmed that this will prevent duplicate queries (you can see this with just 2 labs but it won't start having an impact on performance until you have more than 2 labs)

Reply:
yes, I saw that too in the PyCharm console and was puzzled! Thanks, and sorry, that should have been an obvious place to use get_objects_for_user.